File: challenge-2/main.ts
class MiniCompiler {
  private numericValue: number = 0;
  private output: string = "";

  constructor(program: string) {
    this.compile(program)
  }

  private compile(program: string): void {
    for (const symbol of program) {
      if (symbol === "#") {
        this.numericValue += 1;
      } else if (symbol === "@") {
        this.numericValue -= 1;
      } else if (symbol === "*") {
        this.numericValue *= this.numericValue;
      } else if (symbol === "&") {
        this.output += this.numericValue.toString();
      }
    }
  }

  public getOut(): string {
    return this.output;
  }
}

const solution = new MiniCompiler("&###@&*&###@@##@##&######@@#####@#@#@#@##@@@@@@@@@@@@@@@*&&@@@@@@@@@####@@@@@@@@@#########&#&##@@##@@##@@##@@##@@##@@##@@##@@##@@##@@##@@##@@##@@##@@##@@&");
console.log(solution.getOut())

File: challenge-2/main.py
def mini_compiler(input_str: str) -> str:
    numeric_value = 0
    output = []

    for symbol in input_str:
        if symbol == "#":
            numeric_value += 1
        elif symbol == "@":
            numeric_value -= 1
        elif symbol == "*":
            numeric_value *= numeric_value
        elif symbol == "&":
            output.append(str(numeric_value))

    return ''.join(output)

input_str = "&###@&*&###@@##@##&######@@#####@#@#@#@##@@@@@@@@@@@@@@@*&&@@@@@@@@@####@@@@@@@@@#########&#&##@@##@@##@@##@@##@@##@@##@@##@@##@@##@@##@@##@@##@@##@@##@@&"
output_str = mini_compiler(input_str)
print(output_str)